edgedb/asyncio_con.py

```diff
@@ -241,12 +242,20 @@ async def ensure_connected(self, *, single_attempt=False):
     # overriden by connection pool
     async def _reconnect(self, single_attempt=False):
         inner = self._inner
+        impl = inner._impl
         inner._impl = _AsyncIOConnectionImpl(
             inner._codecs_registry, inner._query_cache)
-        await inner._impl.connect(inner._loop, inner._addrs,
-                                  inner._config, inner._params,
-                                  single_attempt=single_attempt,
-                                  connection=self)
+        try:
+            await inner._impl.connect(inner._loop, inner._addrs,
+                                      inner._config, inner._params,
+                                      single_attempt=single_attempt,
+                                      connection=self)
+        except Exception:
+            # if the new impl fails to connect it will be in an inconsistent
+            # state. Use the old impl instead because it is in a consistent
+            # state.
+            self._inner._impl = impl
```

Comment:
I'm not sure this is the best way to solve this :confused:

Reply:
Call `_AsyncIOConnectionImpl.terminate()` and make sure the latter sets `_protocol` to `None`, which would make `is_closed()` return `True` and there should be no inconsistency downstream.

Reply:
changed